File: 2025-03-08/Vertical_Order_Traversal_Of_A_Binary_Tree.java
import java.util.*;

public class Vertical_Order_Traversal_Of_A_Binary_Tree {
    public static List<List<Integer>> verticalTraversal(TreeNode root) {
        if(root == null) return new ArrayList<>();

        Map<Integer, ArrayList<int[]>> map = new TreeMap<>();
        Queue<Object[]> q = new LinkedList<>();
        q.offer(new Object[]{root, 0});

        while(!q.isEmpty()) {
            Object[] obj = q.poll();
            TreeNode node = (TreeNode) obj[0];
            int level = (int) obj[1];

            map.computeIfAbsent(level, A -> new ArrayList<>()).add(new int[]{node.val, level});

            if(node.left != null)
                q.offer(new Object[]{node.left, level-1});

            if(node.right != null)
                q.offer(new Object[]{node.right, level+1});
        }

        List<List<Integer>> ans = new ArrayList<>();

        for(ArrayList<int[]> arr : map.values()) {
            arr.sort((a, b) -> a[1] == b[1] ? Integer.compare(a[0], b[0]) : Integer.compare(a[1], b[1]));
            ans.add(arr.stream().map(k -> k[0]).toList());
        }

        return ans;
    }
    public static void main(String[] args) {
        Integer[] arr = new Integer[]{3,1,4,0,2,2};
        Tree tree = new Tree();
        TreeNode root = tree.arrayToTree(arr);
        List<List<Integer>> ans = verticalTraversal(root);
        System.out.println(ans.toString());
    }
}

File: 2025-10-09/Vertical_Order_Traversal_Of_A_Binary_Tree.py
from typing import Optional, List

from TreeNode import TreeNode
from Tree import Tree


class Solution:
    def verticalTraversal(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []

        lst = []

        def dfs(root, depth, order):
            if not root:
                return

            lst.append((order, depth, root.val))
            dfs(root.left, depth + 1, order - 1)
            dfs(root.right, depth + 1, order + 1)

        dfs(root, 0, 0)

        lst.sort()
        d = {}
        for order, depth, val in lst:
            if order not in d:
                d[order] = []

            d[order].append(val)

        return list(d.values())


sol = Solution()
tree = Tree()
node = tree.array_to_tree([8, 2, 3, 4, 6, 5, 7])
print(sol.verticalTraversal(node))
